### PR TITLE
Support unrestricted user in CWS

### PR DESCRIPTION
**Pull request by myungwoo**
Support non-restricted user in CWS which discussed in #597 .

---
This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/602)

### CODE REVIEW

File: cms/server/util.py [hunk 1]
```diff
@@ -191,7 +191,8 @@ def actual_phase_required(*actual_phases):
     def decorator(func):
         @wraps(func)
         def wrapped(self, *args, **kwargs):
-            if self.r_params["actual_phase"] not in actual_phases:
+            if self.r_params["actual_phase"] not in actual_phases and \
+                (self.current_user is None or not self.current_user.is_super_user):
```
Comment (stefano-maggiolo) [inline cms/server/util.py]: Codacy Issue found: [Line too long (83/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2394109874/issues/source?bid=3167167&fileBranchId=3230441#l195)

**Comment by stefano-maggiolo** [comment]
Looks good, just a few things:
- when you wrap lines on an "if", the wrapped line should have more indentation, like this:
```python
def fn():
    if is_very_long_line and \
            yes_it_is:
        do_something() 
```
otherwise it will visually merge with the content of the if;
- can you call it "is_unrestricted_user? superuser seems something different;
- there is a line longer than 80 chars (see codacy report).

**Comment by myungwoo** [comment]
I prefer call it "unrestricted" to "is_unrestricted_user" because field name for hidden user is "hidden" !

Thank you for your kind and fast comment.
All you told me is resolved.

**Comment by stefano-maggiolo** [comment]
Ok for unrestricted, will merge later. Thank you!

**Comment by myungwoo** [comment]
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[cms/server/util.py, line 195 \[r1\]](https://reviewable.io:443/reviews/cms-dev/cms/602#-KFAD3SSaFOHx70w--rq-r1-195:-KFAKyvLo2SVNEDmvfdj:-499146009) ([raw file](https://github.com/cms-dev/cms/blob/e252603569811686fd0e4886a2ad4e0e88789b34/cms/server/util.py#L195)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/602)*

File: cms/server/util.py [hunk 2]
```diff
@@ -191,7 +191,9 @@ def actual_phase_required(*actual_phases):
     def decorator(func):
         @wraps(func)
         def wrapped(self, *args, **kwargs):
-            if self.r_params["actual_phase"] not in actual_phases:
+            if self.r_params["actual_phase"] not in actual_phases and \
+                    (self.current_user is None or \
```
Comment (akmohtashami) [inline cms/server/util.py]: shouldn't this be self.current_user is not None?

**Comment by myungwoo** [comment]
**UPD 1**
image
I add these two content.
& update script to remove user (it was a old script which is made before the participation model)
& make script to remove participation
& update script to add participation (add option for unrestricted)

File: cmscontrib/RemoveParticipation.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def remove_participation(contest_id, username):
+    with SessionGen() as session:
+        user = session.query(User)\
+            .filter(User.username == username).first()
+        if user is None:
+            logger.error("User %s does not exist.", username)
+            return False
+
```
Comment (stefano-maggiolo) [inline cmscontrib/RemoveParticipation.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/s-maggiolo/cms/file/2398038415/issues/source?bid=3167167&fileBranchId=3230441#l46)

File: cmscontrib/AddParticipation.py [hunk 4]
```diff
@@ -119,7 +123,7 @@ def main():
 
     return add_participation(args.username, args.contest_id,
                              args.ip, args.delay_time, args.extra_time,
-                             args.password, args.team, args.hidden)
```
Comment (stefano-maggiolo) [inline cmscontrib/AddParticipation.py]: Codacy Issue found: [Line too long (86/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2398038397/issues/source?bid=3167167&fileBranchId=3230441#l126)

Comment (myungwoo) [inline cmscontrib/AddParticipation.py]: Done.

File: cmscontrib/RemoveParticipation.py (hunk 3 above)
Comment (myungwoo) [inline cmscontrib/RemoveParticipation.py]: Done.

**Comment by stefano-maggiolo** [comment]
Done!

**Pull request closed without merging**